### DISCUSSION TITLE
Rewrite the input buffer.

### DISCUSSION THREAD

**Discussion by Windblade-GR01** [ikemen-engine/Ikemen-GO#862] [Engine development]
Currently the input buffer present some problems that make it harder to execute commands.
This is show on these cases like #116.
And any improvements like ones done by NeatUnsou break the functionality of current one.

So to solve problems and allow improvements in the future a full command buffer rewrite is needed.
As compatibly with MUGEN is pority this new buffer would be fully backwards compatible with MUGEN chars.

**Comment by Orochikyocr** [comment]
I closed the issue about keeping 0.94-0.97 input parser as an optional. Once Gacel releases his own input parser it would be good to have an option on the config.json to choose between the two, while Gacel one will be the one turned on by default.

**Comment by Isakku** [comment]
I have a similar problem as in https://github.com/ikemen-engine/Ikemen-GO/issues/116 with my chars
Except it's the opposite problem. When I press forward or backwards, I always get the same result as if I had pressed FF (double forwards) or BB (double backwards). It's really annoying me, I can't just simply walk, I'm always dashing!

EDIT
I must mention I just tried again on WIndows and found out this only happens to me on Linux! Inconsistent behavior huh?

**Comment by assemblaj** [comment]
For what it's worth,  [here's ](https://gist.github.com/assemblaj/494bd3b2bfc5cec329d0b294c00c1f24) a 1:1 port of Turtlewax that I wrote in Go.  I'm not sure how one could integrate this into our current input system but it could be used for inspiration by anyone who'd like to pick this project up.

**Comment by potsmugen** [comment]
I think the problem with referencing a custom buffer is that they approach the problem in a different way. They're coded at the character level in a "if cmd file command is performed, set variable" logic. What the Ikemen parser needs to do (well, already does) is compare the player inputs to the list of commands that are defined in the cmd file.
Not that they can't be used as inspiration as you said.

After Plasmoid Thunder pointed out #1280 to me, I think the current parser will need _at least_ a rewrite of the "ReadCommand" function. The problem of reading inputs was approached in the wrong way from the getgo so some parts of the code aren't salvageable if one wants to make a parser like Mugen's.

I'd be happy to help with this but my knowledge of Golang is very basic and this needs to be tackled by someone with a certain level of mastery over it. I'd still like to help with testing and brainstorming though.

**Comment by potsmugen** [comment]
So I've been working on the input code lately. So far I was able to fix enough bugs to be somewhat confident in saying that it won't need to be rewritten. The bugs that it has are fixable (most are fixed already), and the main issue it had in terms of expandability was being a confusing code. Suehiro understandably wrote the code for himself so it has very little annotations and such. But it's a great code. We just have to finish what he started.

**Comment by potsmugen** [comment]
I'm submitting an input code revamp in #2606, which everyone is welcome to try if/when it gets merged.
